Elly Vending step3

dely2p opened the pull request:
Vending step3 입니다.

dely2p committed:
remove not use variable #13

필요없는 변수들 삭제

dely2p committed:
change VendingMachine from struct to class #13

구조체였던 VendingMachine을 클래스로 바꿈

dely2p committed:
fix AppDelegate to move VendingMachine variable #13

AppDelegate로 VendingMachine 변수를 옮겼다.

dely2p committed:
add decode, encode in InventoryBox, VnedingMachine to archive #14

각 클래스에 decode, encode method를 추가했다.

dely2p committed:
save vendingMachine instance property in background call back method #14

앱 종료 시점 콜백함수에서 vendingMachine 인스턴스 속성 저장하도록 만듦

dely2p committed:
add UnAchiver in AppDelegate applicationDidBecomeActive() method #15

앱델리게이트에서 앱이 액티브 될 때 값을 UnAchive 하도록 만들었다.

dely2p committed:
add required init, encode method to NSObject, NSCoding protocol #16

상위 클래스 Beverage를 상속받는 음료 클래스 전체에 required init과 encode 메소드를 추가했다.

dely2p committed:
add computed property to use same vendingMachine instance in AppDelegate #17

연산 프로퍼티 사용하여 vendingMachine instance를 수정, 접근할 수 있도록 개선함

dely2p committed:
add unArchiving code, get viewController's vendingMachine in AppDelegate application method #17

AppDelegate application method에 언아카이빙하는 코드와 ViewController에 있는 vendingMachine에 접근하여 가져오는 코드를 만들었다.

dely2p committed:
remove computed property #17
AppDelegate에서 ViewController의 인스턴스를 만들고 이를 사용해 updateData라는 연산 프로퍼티를 만들어서 사용하려 했지만 ViweController 인스턴스 역시 사용시점마다 다르기 때문에 싱글톤으로 만들어야하나 고민하다가 다른방법을 사용하여 이 코드는 삭제함

godrm commented on VendingMachineApp/VendingMachineApp/AppDelegate.swift:
```diff
@@ -31,6 +43,8 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
 
     func applicationWillTerminate(_ application: UIApplication) {
+        let data = NSKeyedArchiver.archivedData(withRootObject: vendingMachine)
+        UserDefaults.standard.set(data, forKey: "vendingMachine")
```
applicationWillResignActive, applicationDidEnterBackground, applicationWillTerminate 여기 세 군데 동작이 어떻게 다를까요?
동일한 코드를 중복해서 모두 구현해놓아야만 할까요?

dely2p replied:
- applicationWillResignActive: 홈 버튼을 눌러서 앱이 활성화에서 비활성화 되었을 때,
- applicationDidEnterBackground: 앱이 백그라운드 상태일 때,
- applicationWillTerminate: 앱이 종료될 때 입니다.
setUserDefault() method를 두어 중복되는 코드를 개선했습니다.

godrm commented on VendingMachineApp/VendingMachineApp/AppDelegate.swift:
```diff
@@ -10,18 +10,30 @@ import UIKit
+            vendingMachine = (NSKeyedUnarchiver.unarchiveObject(with: data) as? VendingMachine)!
+        }
+        if let vc = window?.rootViewController?.childViewControllers.first as? ViewController {
+            vc.vendingMachine = vendingMachine
```
여기서 ViewController를 찾아서 vendingMachine을 꼭 넣어야만 하나요?
ViewController에서도 접근할 수 있지 않던가요?

dely2p replied:
어제 설명해주신 것처럼 computed property를 사용하여 ViewController에서 접근하려 했으나 ViewController 인스턴스가 어제 vendingMachine처럼 데이터 업데이트 한 것을 반영하지 못해서요..ㅠㅠ

dely2p replied:
ViewController를 찾아서 vendingMachine을 넣는 부분을 수정했습니다.

godrm reviewed (commented):
수고하셨습니다. 피드백을 확인하세요.

dely2p committed:
remove make viewController instance code #19
AppDelegate에서 viewController instance를 만들어서 vendingMachine을 넣어주던 코드가 중복코드로 필요없기 때문에 지움

godrm commented:
수고하셨어요. 마무리하고 넘어갑시다. 
항상 객체 인스턴스를 참조하는 방법에 따라서, 또는 순서에 따라서
참조하는 객체가 달라지기 때문에 부작용이 발생할 수 있다는 점을 유의하세요.

그리고 문제가 생겼을 때 문제해결을 위해서 가설을 세우고, 논리적으로 추론해서, 디버거로 증명해보고 확인하는 과정을 반복해보세요. 개발자 작업은 코딩하는 것 만큼이나 디버깅해서 문제해결하는 과정이 중요합니다.
